flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/FileSlotAllocationSnapshotPersistenceService.java

```diff
@@ -0,0 +1,136 @@
+                    String.format(
+                            "Cannot create the slot allocation snapshot directory %s.",
+                            slotAllocationSnapshotDirectory));
+        }
+    }
+
+    @Override
+    public void persistAllocationSnapshot(SlotAllocationSnapshot slotAllocationSnapshot)
+            throws IOException {
+        // Let's try to write the slot allocations on file
+        final File slotAllocationSnapshotFile =
+                slotAllocationFile(slotAllocationSnapshot.getSlotID().getSlotNumber());
+
+        try (ObjectOutputStream oos =
+                new ObjectOutputStream(new FileOutputStream(slotAllocationSnapshotFile))) {
+            oos.writeObject(slotAllocationSnapshot);
+
+            LOG.debug(
+                    "Successfully written allocation state metadata file {} for job {} and allocation {}.",
+                    slotAllocationSnapshotFile.toPath(),
+                    slotAllocationSnapshot.getJobId(),
+                    slotAllocationSnapshot.getAllocationId());
+        }
+    }
+
+    private File slotAllocationFile(int slotIndex) {
+        return new File(
+                slotAllocationSnapshotDirectory.getAbsolutePath(), slotIndexToFilename(slotIndex));
+    }
+
+    private static String slotIndexToFilename(int slotIndex) {
+        return slotIndex + SUFFIX;
+    }
+
+    private static int filenameToSlotIndex(String filename) {
+        return Integer.parseInt(filename.substring(0, filename.length() - SUFFIX.length()));
+    }
+
+    @Override
+    public void deleteAllocationSnapshot(int slotIndex) {
+        // Let's try to write the slot allocations on file
+        final File slotAllocationSnapshotFile = slotAllocationFile(slotIndex);
+        try {
+            FileUtils.deleteFileOrDirectory(slotAllocationSnapshotFile);
+            LOG.debug(
+                    "Successfully deleted allocation state metadata file {}.",
+                    slotAllocationSnapshotFile.toPath());
+        } catch (IOException ioe) {
+            LOG.warn(
+                    "Cannot delete the local allocations state file {}.",
+                    slotAllocationSnapshotFile.toPath(),
+                    ioe);
+        }
+    }
+
+    @Override
+    public Collection<SlotAllocationSnapshot> loadAllocationSnapshots() {
+        // Let's try to populate the slot allocation from local file
+        final File[] slotAllocationFiles = slotAllocationSnapshotDirectory.listFiles();
+        if (slotAllocationFiles == null) {
+            LOG.debug("No allocation files to load.");
+            return Collections.emptyList();
+        }
+
+        Collection<SlotAllocationSnapshot> slotAllocationSnapshots =
+                new ArrayList<>(slotAllocationFiles.length);
+
+        for (File allocationFile : slotAllocationFiles) {
+            try (ObjectInputStream ois =
+                    new ObjectInputStream(new FileInputStream(allocationFile))) {
```

Comment:
Something to keep in mind is that this allows someone with write access to the workingDirectory to execute arbitrary code.
I think we'd want to migrate this to something like jackson soon (i.e., before the release), which should be pretty straight-forward.

Reply:
This is a good point. I will try to open a follow-up PR for it asap.